### ISSUE TITLE
Add support for the "required" option to unix-char/unix-block

### ISSUE THREAD

**Issue by stgraber** [lxc/lxd#3995]
unix-char and unix-block should gain the "required" flag that we have for USB devices.

This would allow for required=false to be set on them, in which case an inotify handler would be set up waiting for the device source path to appear in the device's parent directory (typically /dev). Once it does, then the device would be created inside the container. If the path goes away from the host, then the device should also be removed from the container.

In short:

 - Add "required" as a property to unix-char and unix-block (defaulting to "true" for backward compatibility)
 - Maintain a central inotify handler for devices and make sure we subscribe to notifications for the parent directory of any unix device that we know of.
 - If we notice a removal event, remove it from the containers.
 - If we notice an addition event, look for any running container that wants that device and pass it to it.

**Comment by brauner** [comment]
There's not `inotify` support in official golang. There are a few packages but for our purposes I think it makes sense to simply write a couple of small wrappers ourselves.

**Comment by stgraber** [comment]
I've used `gopkg.in/fsnotify.v0` in the past for this (in lxd-demo-server) and that worked pretty well, but it may be overkill in this case since we don't really need it to work on different OSes.

**Comment by brauner** [comment]
On Jan 5, 2018 04:43, "Stéphane Graber" <notifications@github.com> wrote:

I've used gopkg.in/fsnotify.v0 in the past for this (in lxd-demo-server)
and that worked pretty well, but it may be overkill in this case since we
don't really need it to work on different OSes.

Overkill and also doesn't use the syscall package but the x packages.


—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/lxc/lxd/issues/3995#issuecomment-355465824>, or mute
the thread
